### PR TITLE
Dev identification modularization, speed optimization

### PR DESCRIPTION
**Pull request by CCranney**

### CODE REVIEW

File: csodiaq/csodiaqParser.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+        )
+
+
+def create_output_name(commandName):
+    return f"csodiaq-{commandName}-{time.strftime('%Y%m%d-%H%M%S')}"
+
+
+def create_new_output_directory_path(newDirectoryLocation, commandName):
+    if newDirectoryLocation[-1] == "/":
+        newDirectoryLocation = newDirectoryLocation[:-1]
```
Comment (AlexandreHutton) [inline csodiaq/csodiaqParser.py]: `os.path.join` takes care of this; you can remove these lines.

File: csodiaq/csodiaqParser.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    newDirectoryName = create_output_name(commandName)
+    if os.path.isdir(newDirectoryLocation):
+        newDirectoryPath = os.path.join(newDirectoryLocation, newDirectoryName)
+    else:
+        newDirectoryHeader = newDirectoryLocation.split("/")[-1]
+        newDirectoryParentDirectory = "/".join(newDirectoryLocation.split("/")[:-1])
```
Comment (AlexandreHutton) [inline csodiaq/csodiaqParser.py]: `os.path.dirname`

File: csodiaq/csodiaqParser.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+
+def create_new_output_directory_path(newDirectoryLocation, commandName):
+    if newDirectoryLocation[-1] == "/":
+        newDirectoryLocation = newDirectoryLocation[:-1]
+    newDirectoryName = create_output_name(commandName)
+    if os.path.isdir(newDirectoryLocation):
+        newDirectoryPath = os.path.join(newDirectoryLocation, newDirectoryName)
+    else:
+        newDirectoryHeader = newDirectoryLocation.split("/")[-1]
```
Comment (AlexandreHutton) [inline csodiaq/csodiaqParser.py]: `os.path.basename`

File: csodiaq/csodiaqParser.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        raise argparse.ArgumentTypeError(
+            "The protein argument requires the presence of protein FDR files to function. Please run the protein scoring workflow or remove the protein argument from your commands."
+        )
+
+
+def create_output_name(commandName):
+    return f"csodiaq-{commandName}-{time.strftime('%Y%m%d-%H%M%S')}"
+
+
+def create_new_output_directory_path(newDirectoryLocation, commandName):
```
Comment (AlexandreHutton) [inline csodiaq/csodiaqParser.py]: I would expect `create_new_output_directory_path` to make the directory (i.e., `os.mkdir`). I'd recommend switching to `get_pathname` or something similar.

File: csodiaq/csodiaqParser.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+
+    def __call__(self, newDirectoryLocation):
+        if os.path.isfile(newDirectoryLocation):
+            raise argparse.ArgumentTypeError(
+                "The -o or --output argument must be a directory or to-be-created directory header, not an existing file."
+            )
+        if not os.path.isdir(os.path.dirname(newDirectoryLocation)):
+            raise argparse.ArgumentTypeError(
+                "The -o or --output argument directory requires an existing parent directory."
+            )
```
Comment (AlexandreHutton) [inline csodiaq/csodiaqParser.py:182]: `OutputDirectory` shouldn't have knowledge of the argument list; it just has `commandName` and `newDirectoryLocation` when called. I would remove the `ArgumentTypeError` and replace it with `ValueError`, indicating that the directory already exists. The parser should deal with raised exceptions as appropriate (if you want to report `ArgumentTypeError` to the user, then have the parser catch the ValueError and report.

File: csodiaq/csodiaqParser.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+    def __init__(self):
+        self.allowedFileTypes = ["mzXML"]
+
+    def __call__(self, inputQueryFile):
+        if not os.path.isfile(inputQueryFile):
+            raise argparse.ArgumentTypeError(
+                "The -i or --input argument must be an existing file (and not a directory)."
+            )
+        if inputQueryFile.split(".")[-1] not in self.allowedFileTypes:
+            raise argparse.ArgumentTypeError(
```
Comment (AlexandreHutton) [inline csodiaq/csodiaqParser.py:199]: Same as `OutputDirectory`

File: csodiaq/csodiaqParser.py [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+        self.allowedFileTypes = ["csv", "tsv", "mgf"]
+
+    def __call__(self, libraryFile):
+        if not os.path.isfile(libraryFile):
+            raise argparse.ArgumentTypeError(
+                "The -l or --library argument must be an existing file (and not a directory)."
+            )
+        if libraryFile.split(".")[-1] not in self.allowedFileTypes:
+            raise argparse.ArgumentTypeError(
+                "The -l or --library argument must be a .tsv, .csv or .mgf file."
```
Comment (AlexandreHutton) [inline csodiaq/csodiaqParser.py:216]: Same as `OutputDirectory`

File: csodiaq/csodiaqParser.py [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+            raise argparse.ArgumentTypeError(
+                f"The {self.type} argument must be {self.return_expected_type()}."
+            )
+        if value < self.minValue:
+            raise argparse.ArgumentTypeError(
+                f"The {self.type} argument must be {self.return_expected_type()} greater than or equal to {self.minValue}."
+            )
+        if value > self.maxValue:
+            raise argparse.ArgumentTypeError(
+                f"The {self.type} argument must be {self.return_expected_type()} less than or equal to {self.maxValue}."
```
Comment (AlexandreHutton) [inline csodiaq/csodiaqParser.py:249]: Same as `OutputDirectory`

File: csodiaq/csodiaqParser.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+        return "a float"
+
+    def coerce_into_expected_type(self, value):
+        return float(value)
+
+    def __call__(self, value):
+        value = super().__call__(value)
+        if self.type == "binValueProximity" and not round(value, 2) == value:
+            raise argparse.ArgumentTypeError(
+                f"The {self.type} argument cannot have values beyond 2 decimal places (mass spectrometers are typically not sensitive enough for that specificity)."
```
Comment (AlexandreHutton) [inline csodiaq/csodiaqParser.py:273]: Same as `OutputDirectory`

File: csodiaq/csodiaqParser.py [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+            )
+        return value
+
+
+class CsodiaqOutputDirectory(ABC):
+    def __call__(self, idDir):
+        if not os.path.isdir(idDir):
+            raise argparse.ArgumentTypeError(
+                "The -i or --input argument must be a directory."
```
Comment (AlexandreHutton) [inline csodiaq/csodiaqParser.py:282]: Same as `OutputDirectory`

File: csodiaq/csodiaqParser.py (hunk 5 above)
Comment (AlexandreHutton) [reply csodiaq/csodiaqParser.py:182]: Alternatively, use the leading underscore to indicate that the class is not meant to be used outside of this context.

**Review by AlexandreHutton — CHANGES_REQUESTED** [review]
Minor suggestions, mostly with path handling. Looks good otherwise.

File: csodiaq/csodiaqParser.py (hunk 1 above)
Comment (CCranney) [reply csodiaq/csodiaqParser.py]: The later suggestions fix the concern I had here - part of the issue was that I needed to create a new directory name if it didn't exist, and the slash could have gotten in the way of the str.split() function. But with using os.path.dirname() and os.path.basename() this concern is eliminated, thank you!

File: csodiaq/csodiaqParser.py (hunk 2 above)
Comment (CCranney) [reply csodiaq/csodiaqParser.py]: updated

File: csodiaq/csodiaqParser.py (hunk 3 above)
Comment (CCranney) [reply csodiaq/csodiaqParser.py]: updated

File: csodiaq/csodiaqParser.py (hunk 4 above)
Comment (CCranney) [reply csodiaq/csodiaqParser.py]: get point, switched "create" to "get" in each case

File: csodiaq/csodiaqParser.py (hunk 5 above)
Comment (CCranney) [reply csodiaq/csodiaqParser.py:182]: I'm a little confused about the issue here. ArgumentTypeError is an argparse-specific error, and using it returns the provided message to the end user, indicating why the parsing failed. Raising a ValueError in an argparse.parse_args() call fails to show the error message, and instead says that the class object failed generally, as below:

### with ArgumentTypeError:
```
(venv) (base) cranneyc@awlpdcg556mm unit % csodiaq id -i <path>/DELETE.mzXML -l <path>/library_files -o <path>/outputDir -hist
usage: csodiaq id [-h] -o OUTPUT -i INPUT -l LIBRARY [-t MATCHTOLERANCE] [-nc] [-c CORRECTIONDEGREE] [-hist]
csodiaq id: error: argument -l/--library: The -l or --library argument must be an existing file (and not a directory).
```

### with ValueError:
```
(venv) (base) cranneyc@awlpdcg556mm unit % csodiaq id -i <path>/DELETE.mzXML -l <path>/library_files -o <path>/outputDir
usage: csodiaq id [-h] -o OUTPUT -i INPUT -l LIBRARY [-t MATCHTOLERANCE] [-nc] [-c CORRECTIONDEGREE] [-hist]
csodiaq id: error: argument -l/--library: invalid <csodiaq.csodiaqParser.LibraryFile object at 0x140e18bb0> value: '<path>/library_files'
```

If asking about the use of these classes generally, they are also used by the GUI to evaluate which fields were filled out incorrectly, highlighting offending ones as red and passing ones as green. I figure packaging up related errors attached to single fields into one class is a concise way to use them, both here and in the GUI.

Am I misunderstanding the issue?

Comment (AlexandreHutton) [reply csodiaq/csodiaqParser.py:182]:
```Python
import os
from csodiaqParser import OutputDirectory
out_dir = OutputDirectory("command")
os.mkdir("conflicting_dir")

out_dir("conflicting_dir")
```
That should raise ArgumentTypeError, informing the user that they used a wrong value for the -o argument, even though the command line was never used and argparse was not called. The error message is giving the wrong information about what went wrong; it's not that the `-o` parameter is wrong, it's that the file/directory already exists.
You can keep it the error as-is, but you should indicate that the class shouldn't be used outside of its current context (leading underscore `_OutputDirectory` vs `OutputDirectory`). Similarly for the others.

Comment (CCranney) [reply csodiaq/csodiaqParser.py:182]: Got it - it's purely about specifying to the end user that this is a class that should only be used in an argparse setting. Sounds good, I'll add underscores to reflect that. Thanks!

**Pull request merged**